### ISSUE TITLE
patchelf: section header table out of bounds

### ISSUE THREAD

**Issue by asavah** [NixOS/patchelf#263]
**Describe the bug**

```
patchelf --debug --set-rpath /home/asavah/kross/host/lib anyelffile
patching ELF file 'anyelffile'
patchelf: section header table out of bounds
```

**Steps To Reproduce**

see above

**Expected behavior**

rpath changed without error

patchelf built from git at https://github.com/NixOS/patchelf/commit/0ad9548c02218ac1e6d0dc1a96d0bc955ea686ee

**Additional context**

It looks like https://github.com/NixOS/patchelf/commit/cedf28411112fef1b2fc0ac848d88991efa06c87 messed up the parenthesis and changed the logic

before:

`(rdi(hdr->e_shoff) + rdi(hdr->e_shnum) * rdi(hdr->e_shentsize)) > fileContents->size()`
A+B*C > D

after:

`(rdi(hdr->e_shoff) + rdi(hdr->e_shnum)) * rdi(hdr->e_shentsize) > fileContents->size()`
(A+B)*C > D

The following patch seems to restore the logic and fix the issue here:
```
diff --git a/src/patchelf.cc b/src/patchelf.cc
index 50eb13a..9b75b6a 100644
--- a/src/patchelf.cc
+++ b/src/patchelf.cc
@@ -397,13 +397,13 @@ ElfFile<ElfFileParamNames>::ElfFile(FileContents fileContents)
     if (rdi(hdr->e_type) != ET_EXEC && rdi(hdr->e_type) != ET_DYN)
         error("wrong ELF type");

-    if (size_t(rdi(hdr->e_phoff) + rdi(hdr->e_phnum)) * rdi(hdr->e_phentsize) > fileContents->size())
+    if (size_t(rdi(hdr->e_phoff) + rdi(hdr->e_phnum) * rdi(hdr->e_phentsize)) > fileContents->size())
         error("program header table out of bounds");

     if (rdi(hdr->e_shnum) == 0)
         error("no section headers. The input file is probably a statically linked, self-decompressing binary");

-    if (size_t(rdi(hdr->e_shoff) + rdi(hdr->e_shnum)) * rdi(hdr->e_shentsize) > fileContents->size())
+    if (size_t(rdi(hdr->e_shoff) + rdi(hdr->e_shnum) * rdi(hdr->e_shentsize)) > fileContents->size())
         error("section header table out of bounds");

     if (rdi(hdr->e_phentsize) != sizeof(Elf_Phdr))
```

With the patch:
```
patchelf --set-rpath /home/asavah/kross/host/lib mkcodecache
patching ELF file 'mkcodecache'

readelf -d mkcodecache | grep RUNP
0x000000000000001d (RUNPATH)            Library runpath: [/home/asavah/kross/host/lib]
```

**Comment by domenkozar** [comment]
cc @neheb

**Comment by neheb** [comment]
This is the error: https://clang.llvm.org/extra/clang-tidy/checks/bugprone-misplaced-widening-cast.html

It sounds like it should be (B * C) + A or (C * B) + A

**Comment by digisomni** [comment]
This has also been problematic for me in https://github.com/vircadia/vircadia-builder so I'm going to checkout the parent of that breaking commit for the while. :)

**Comment by asavah** [comment]
fixed by #266
